bblocks/import_tools/weo.py

```diff
@@ -0,0 +1,287 @@
+"""Tools to import IMF World Economic Outlook data."""
+
+import xml.etree.ElementTree as ET
+from zipfile import ZipFile
+from dataclasses import dataclass
+import pandas as pd
+from datetime import datetime
+from bs4 import BeautifulSoup
+import io
+
+from bblocks.import_tools.common import ImportData, get_response, unzip
+from bblocks.config import BBPaths
+from bblocks.logger import logger
+
+BASE_URL = "https://www.imf.org/"
+
+COLUMN_MAPPER = {
+    "UNIT": "IMF.CL_WEO_UNIT.1.0",
+    "CONCEPT": "IMF.CL_WEO_CONCEPT.1.0",
+    "REF_AREA": "IMF.CL_WEO_REF_AREA.1.0",
+    "FREQ": "IMF.CL_FREQ.1.0",
+    "SCALE": "IMF.CL_WEO_SCALE.1.0",
+}
+
+
+@dataclass
+class Parser:
+    """Helper class to download WEO"""
+
+    folder: ZipFile
+
+    data_file = None
+    schema_file = None
+    data = None
+
+    def get_files(self):
+        """Get the root of files"""
+
+        if len(self.folder.namelist()) > 2:
+            raise ValueError("More than one file in zip file")
+
+        for file in self.folder.namelist():
+            if file.endswith(".xml"):
+                self.data_file = ET.parse(self.folder.open(file)).getroot()
+
+            if file.endswith(".xsd"):
+                self.schema_file = ET.parse(self.folder.open(file)).getroot()
+
+        if self.data_file is None or self.schema_file is None:
+            raise ValueError("No data or schema file found")
+
+    def extract_data(self):
+        """extract data from data file"""
+
+        rows = []
+        for series in self.data_file[1].findall("./"):
+            for obs in series.findall("./"):
+                rows.append({**series.attrib, **obs.attrib})
+
+        self.data = pd.DataFrame(rows)
+
+    def _convert_series_codes(self, series: pd.Series, lookup_value: str):
+        """ """
+
+        query = self.schema_file.findall(
+            "./{http://www.w3.org/2001/XMLSchema}simpleType[@name="
+            + f'"{lookup_value}"'
+            + "]/"
+        )[0].findall("./")
+
+        return series.map({elem.attrib["value"]: elem[0][0].text for elem in query})
+
+    def clean_data(self):
+        """Format the dataframe"""
+
+        for col, value in COLUMN_MAPPER.items():
+            self.data = self.data.rename(columns={col: f"{col}_CODE"})
+            self.data[col] = self._convert_series_codes(self.data[f"{col}_CODE"], value)
+
+        for col in ["REF_AREA_CODE", "LASTACTUALDATE", "TIME_PERIOD", "OBS_VALUE"]:
+            self.data[col] = pd.to_numeric(self.data[col], errors="coerce")
+
+    def parse_data(self) -> pd.DataFrame:
+        """Parse the data to a dataframe"""
+
+        self.get_files()
+        self.extract_data()
+        self.clean_data()
+
+        return self.data
```

Comment:
I think you should write a 'getter' for the data. Otherwise the only 'method' to get the data is to run through the pipeline (or access it directly, which isn't great). `parse_data` could still return the data, as well as setting it, but it would be good to have a separate way to get to the data directly (if the pipeline has already been run).